### PR TITLE
Fixed #18707 -- Added support for the test client to return 500 responses.

### PR DESCRIPTION
**Pull request by jdufresne**
https://code.djangoproject.com/ticket/18707

### CODE REVIEW

File: django/test/client.py [hunk 1]
```diff
@@ -440,9 +440,10 @@ class Client(RequestFactory):
     contexts and templates produced by a view, rather than the
     HTML rendered to the end-user.
     """
-    def __init__(self, enforce_csrf_checks=False, **defaults):
+    def __init__(self, enforce_csrf_checks=False, raise_view_exception=True, **defaults):
```
[the remark this answers is not in the document]
Comment (jdufresne) [reply django/test/client.py]: What do you think about `raise_request_exception`?

**Comment by carltongibson** [comment]
Hey @jdufresne. First glance this seems very exciting! I need to get it into a project to test but assuming it does what it says, super! 

Reminiscent of [Trac #29642](https://code.djangoproject.com/ticket/29642)
Would make testing error views a lot easier than using a custom middleware: 

https://github.com/django/django/blob/7e978fdc4228eb44cf97cb4243adc7b7bfd586c7/tests/handlers/tests_custom_error_handlers.py#L35

File: django/test/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (charettes) [reply django/test/client.py]: I like it!

**Comment by jdufresne** [comment]
I renamed `raise_view_exception` to `raise_request_exception` and tweaked the docs to make it more clear that it affects any exception from the request.

File: docs/topics/testing/tools.txt [hunk 2]
```diff
@@ -128,6 +128,14 @@ Use the ``django.test.Client`` class to make requests.
     The ``json_encoder`` argument allows setting a custom JSON encoder for
     the JSON serialization that's described in :meth:`post`.
 
+    The ``raise_request_exception`` argument allows controlling whether or not
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Modify the signature above.

File: docs/releases/3.0.txt [hunk 3]
```diff
@@ -185,7 +185,13 @@ Templates
 Tests
 ~~~~~
 
-* ...
+* The new test :class:`~django.test.Client` argument
```
Comment (timgraham) [inline docs/releases/3.0.txt]: Normally I keep the release note simpler (1 sentence) and link to the documentation that provides a more information. In this case the release note seems more detailed than the documentation which is backwards.

Also, I think the documentation could better describe the use case.

File: docs/topics/testing/tools.txt [hunk 4]
```diff
@@ -555,6 +581,11 @@ The only exceptions that are not visible to the test client are
 exceptions internally and converts them into the appropriate HTTP response
 codes. In these cases, you can check ``response.status_code`` in your test.
 
+If ``Client.raise_request_exception`` is ``False``, the test client will return a
+500 response as would be returned to a browser. The response has the attribute
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Mention/link to `handler500`?

File: docs/topics/testing/tools.txt [hunk 5]
```diff
@@ -555,6 +581,11 @@ The only exceptions that are not visible to the test client are
...
+:attr:`~Response.exc_info` to provide information about the unhandled
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Normally the name of the attribute precedes the word attribute, i.e. "exc_info attribute".

File: tests/test_client/tests.py [hunk 6]
```diff
@@ -759,6 +759,20 @@ def test_view_with_exception(self):
         with self.assertRaises(KeyError):
             self.client.get("/broken_view/")
 
+    def test_exc_info(self):
+        client = Client(raise_request_exception=False)
+        response = client.get("/broken_view/")
```
Comment (timgraham) [inline tests/test_client/tests.py]: Don't bother changing now but please use single quotes  in the future.

**Pull request merged**